src/java.base/share/classes/java/lang/ModuleLayer.java

```diff
@@ -297,6 +299,36 @@ public Controller addOpens(Module source, String pn, Module target) {
             source.implAddOpens(pn, target);
             return this;
         }
+
+        /***
+         * Enables native access for a module in the layer if the caller's module
+         * has native access.
+         *
+         * <p> This method is <a href="foreign/package-summary.html#restricted"><em>restricted</em></a>.
+         * Restricted methods are unsafe, and, if used incorrectly, their use might crash
+         * the JVM or, worse, silently result in memory corruption. Thus, clients should refrain
+         * from depending on restricted methods, and use safe and supported functionalities,
+         * where possible.
+         *
+         * @param  target
+         *         The module to update
+         *
+         * @return This controller
+         *
+         * @throws IllegalCallerException if access to this method occurs from a module {@code M}
+         *         and the command line option {@code --enable-native-access} is specified, but
+         *         does not mention the module name {@code M}, or {@code ALL-UNNAMED} in case
+         *         {@code M} is an unnamed module.
+         */
```

Comment:
The implementation changes are okay but I think the javadoc, particularly the description for IllegalCallerException, will need a bit of word smithing as we can't use informal language like "does not mention the module name" or talk about command line options or operands that the RI provides. The method description links to the "Restricted methods" section in the package description and maybe lend on that rather than trying to repeat in the exception description. It may be that you can just say that the caller is in a module that does not have native access enabled. That will work for module layers because the caller may had its native access enabled via the API rather than the command line.

The method will also needs a @throws IllegalArgumentException for when the module is not in the module layer.

Reply:
I agree with the comment - e.g. not specifying command line options in the API javadoc. However this is the standard text we have on all restricted methods, so for now I think this change seems at least consistent with all the other restricted methods, and we'll need to do another pass to fix them all.

Reply:
javadoc blurb on IllegalCallerException is from other similar Panama APIs. We need to revisit all those (perhaps as separate patch). I've fixed that and added IllegalArgumentException as you suggested.